pkg/conformance_testing/messages.go

```diff
@@ -0,0 +1,159 @@
+package conformance_testing
+
+import (
+	"fmt"
+	"github.com/eigerco/strawberry/internal/block"
+	"github.com/eigerco/strawberry/internal/crypto"
+	"github.com/eigerco/strawberry/internal/state/serialization/statekey"
+	"github.com/eigerco/strawberry/pkg/serialization/codec/jam"
+	"io"
+)
+
+const (
+	peerInfoKind    = 0
+	importBlockKind = 1
+	setStateKind    = 2
+	getStateKind    = 3
+	stateKind       = 4
+	stateRootKind   = 5
+)
+
+type MessageChoice interface {
+	isMessageChoice()
+}
+
+type Version struct {
+	Major int // 0..255
+	Minor int // 0..255
+	Patch int // 0..255
+}
+
+func (v Version) String() string {
+	return fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Patch)
+}
+
+type PeerInfo struct {
+	Name       []byte
+	AppVersion Version
+	JamVersion Version
+}
+
+func (p PeerInfo) isMessageChoice() {}
+
+type SetState struct {
+	Header block.Header
+	State  State
+}
+
+func (s SetState) isMessageChoice() {}
+
+type StateRoot struct {
+	StateRootHash crypto.Hash
+}
+
+func (s StateRoot) isMessageChoice() {}
+
+type KeyValue struct {
+	Key   statekey.StateKey
+	Value []byte
+}
+
+type ImportBlock struct {
+	Block block.Block
+}
+
+func (i ImportBlock) isMessageChoice() {}
+
+type GetState struct {
+	HeaderHash crypto.Hash
+}
+
+func (g GetState) isMessageChoice() {}
+
+type State struct {
+	StateItems []KeyValue
+}
+
+func (s State) isMessageChoice() {}
+
+type Message struct {
+	// One of: PeerInfo, ImportBlock, SetState, GetState, State, StateRoot
+	Choice MessageChoice
+}
+
+func (m *Message) UnmarshalJAM(reader io.Reader) error {
+	msgType := make([]byte, 1)
+	if _, err := reader.Read(msgType); err != nil {
+		return err
+	}
+	switch msgType[0] {
+	case peerInfoKind:
+		innerMsg := PeerInfo{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	case importBlockKind:
+		innerMsg := ImportBlock{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	case setStateKind:
+		innerMsg := SetState{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	case getStateKind:
+		innerMsg := GetState{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	case stateKind:
+		innerMsg := State{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	case stateRootKind:
+		innerMsg := StateRoot{}
+		if err := jam.NewDecoder(reader).Decode(&innerMsg); err != nil {
+			return err
+		}
+		m.Choice = innerMsg
+	default:
+		return fmt.Errorf("unknown choice type")
+	}
+	return nil
+}
+
+func (m *Message) MarshalJAM() ([]byte, error) {
```

Comment:
This will work for now, but I think longer term we might want to refactor this to be similar to how we normally handle enums, ie [here](https://github.com/eigerco/strawberry/blob/main/internal/safrole/safrole.go#L20), using [this interface](https://github.com/eigerco/strawberry/blob/main/pkg/serialization/codec/jam/types.go#L9).